phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/db/cloud/CloudType.kt

```diff
@@ -0,0 +1,116 @@
+//
+// We make every attempt to re-use code from the database serialization routines.
+// However, cloud serialization is a bit different from database serialization.
+//
+// DIFFERENCE #1:
+//
+// SqlDelight allows us to version the database, and make changes to the database structure.
+// For example, when upgrading the app from v1.5 to v1.6,
+// we might upgrade the database from v2 to v3.
+// This upgrade mechanism allows us to write code that **only** supports database v3,
+// since SqlDelight handles the upgrade mechanics.
+//
+// This upgrade mechanism isn't available for the cloud.
+// Thus, in the cloud we might have serialized objects in v1, v2, v3...
+// And we will need to support deserializing all these versions.
+//
+// DIFFERENCE #2:
+//
+// For the local system, space is cheap, and the disk is fast.
+// But for the cloud, space is really expensive.
+// On iOS, the user only has 5 GB. But that's NOT per app.
+// That 5GB is meant to be shared by every app on their phone.
+// Which means we're expected to be a good steward of their cloud space.
+//
+// So we make certain changes.
+//
+// OPTIMIZATION #1:
+//
+// We use CBOR instead of JSON.
+// This allows us to very efficiently encode ByteArray's.
+//
+// For a discussion of the space savings in practice,
+// see the comments in CloudSerializers.kt.
+//
+// OPTIMIZATION #2:
+//
+// UUID is serialized as: {
+//   mostSignificantBits: Long,
+//   leastSignificantBits: Long
+// }
+// This is improved using a custom serializer.
+//
+
+@Serializable
+data class CloudData @OptIn(ExperimentalSerializationApi::class) constructor(
+    @ByteString
+    @SerialName("i")
+    val incoming: IncomingPaymentWrapper?,
+    @ByteString
+    @SerialName("o")
+    val outgoing: OutgoingPaymentWrapper?,
+    @SerialName("v")
+    val version: Int,
+    @ByteString
+    @SerialName("p")
+    val padding: ByteArray?,
+) {
+    constructor(incoming: IncomingPayment, version: Int) : this(
+        incoming = IncomingPaymentWrapper(incoming),
+        outgoing = null,
+        version = version,
+        padding = ByteArray(size = 0)
+    )
+
+    constructor(outgoing: OutgoingPayment, version: Int) : this(
+        incoming = null,
+        outgoing = OutgoingPaymentWrapper(outgoing),
+        version = version,
+        padding = ByteArray(size = 0)
+    )
+
+    // This function exists because the Kotlin-generated
+    // copy function doesn't translate to iOS very well.
+    //
+    fun copyWithPadding(padding: ByteArray): CloudData {
+       return this.copy(padding = padding)
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+fun CloudData.cborSerialize(): ByteArray {
+    return Cbor.encodeToByteArray(this)
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+fun CloudData.Companion.cborDeserialize(blob: ByteArray): CloudData? {
+    var result: CloudData? = null
+    try {
+        result = Cbor.decodeFromByteArray(blob)
+    } catch (e: Throwable) {}
+
+    return result?.freeze()
```

Comment:
Why do we need to freeze the `CloudData` object? If it's important on iOS, we could to add a test in `iosTest`.

Reply:
Good question. After downloading a CloudData blob, we have to:
- decrypt the blob
- then deserialize the cleartext (into a CloudData instance)

Both of these operations can be considered CPU intensive, so they're performed on a background thread. But Kotlin Multiplatform has horrible support for background threading. (I would go so far as to say that KMM has anti-support for background threading.)

So we try to decrypt+deserialize on a background thread, and then pass the CloudData instance to the main thread. But unfortunately KMM has already tagged the CloudData instance as "contaminated", since it touched a background thread. And so it crashes ! :man_facepalming:

The workaround is to freeze the object, so it's "safe" for use on the main thread.

Perhaps the freezing process shouldn't be built into the deserialization method. But rather, is an additional function that can be explicitly called from Swift.

Reply:
Done in 9d41177

I didn't add a unit test in Kotlin because I'm not sure it makes sense anymore. The Swift layer is now responsible for calling `freeze`, since this is an iOS quirk.